test/utils/oidc/testserver.go

```diff
@@ -0,0 +1,214 @@
+)
+
+type TestServer struct {
+	httpServer   *httptest.Server
+	tokenHandler *MockTokenHandler
+	jwksHandler  *MockJWKsHandler
+	TearDownFn   func()
+}
+
+func (ts *TestServer) JwksHandler() *MockJWKsHandler {
+	return ts.jwksHandler
+}
+
+func (ts *TestServer) TokenHandler() *MockTokenHandler {
+	return ts.tokenHandler
+}
+
+func (ts *TestServer) URL() string {
+	return ts.httpServer.URL
+}
+
+func (ts *TestServer) TokenURL() string {
+	return ts.httpServer.URL + tokenWebPath
+}
+
+func BuildAndRunTestServer(t *testing.T, caPath, caKeyPath string) *TestServer {
+	t.Helper()
+
+	certContent, err := os.ReadFile(caPath)
+	require.NoError(t, err)
+	keyContent, err := os.ReadFile(caKeyPath)
+	require.NoError(t, err)
+
+	cert, err := tls.X509KeyPair(certContent, keyContent)
+	require.NoError(t, err)
+
+	mux := http.NewServeMux()
+	httpServer := httptest.NewUnstartedServer(mux)
+	httpServer.TLS = &tls.Config{
+		Certificates: []tls.Certificate{cert},
+	}
+	httpServer.StartTLS()
+
+	mockCtrl := gomock.NewController(t)
+
+	oidcServer := &TestServer{
+		httpServer:   httpServer,
+		tokenHandler: NewMockTokenHandler(mockCtrl),
+		jwksHandler:  NewMockJWKsHandler(mockCtrl),
+		TearDownFn: func() {
+			mockCtrl.Finish()
+			httpServer.Close()
+		},
+	}
+
+	mux.HandleFunc(openIDWellKnownWebPath, func(writer http.ResponseWriter, request *http.Request) {
+		err := json.NewEncoder(writer).Encode(struct {
+			Issuer      string `json:"issuer"`
+			AuthURL     string `json:"authorization_endpoint"`
+			TokenURL    string `json:"token_endpoint"`
+			JWKSURL     string `json:"jwks_uri"`
+			UserInfoURL string `json:"userinfo_endpoint"`
+		}{
+			Issuer:      httpServer.URL,
+			AuthURL:     httpServer.URL + authWebPath,
+			TokenURL:    httpServer.URL + tokenWebPath,
+			JWKSURL:     httpServer.URL + jwksWebPath,
+			UserInfoURL: httpServer.URL + userInfoWebPath,
+		})
+		require.NoError(t, err)
+
+		writer.Header().Add("Content-Type", "application/json")
+		writer.WriteHeader(http.StatusOK)
+
+		return
+	})
+
+	mux.HandleFunc(tokenWebPath, func(writer http.ResponseWriter, request *http.Request) {
+		token, err := oidcServer.tokenHandler.Token()
+		if err != nil {
+			http.Error(writer, err.Error(), http.StatusBadRequest)
+
+			return
+		}
+
+		writer.Header().Add("Content-Type", "application/json")
+		writer.WriteHeader(http.StatusOK)
+
+		err = json.NewEncoder(writer).Encode(token)
+		require.NoError(t, err)
+
+		return
+	})
+
+	mux.HandleFunc(authWebPath, func(writer http.ResponseWriter, request *http.Request) {
+		writer.WriteHeader(http.StatusOK)
+		return
+	})
+
+	mux.HandleFunc(jwksWebPath, func(writer http.ResponseWriter, request *http.Request) {
+		keySet := oidcServer.jwksHandler.KeySet()
+
+		writer.Header().Add("Content-Type", "application/json")
+		writer.WriteHeader(http.StatusOK)
+
+		err := json.NewEncoder(writer).Encode(keySet)
+		require.NoError(t, err)
+	})
+
+	return oidcServer
+}
+
+func DefaultTokenHandlerBehaviour(
```

Comment:
Given how many arguments this takes it shouldn't be called `DefaultTokenHandlerBehaviour`